File: VoidLink/UIAppView.m
//
//  UIAppView.m
//  Moonlight
//
//  Created by Diego Waxemberg on 10/22/14.
//  Copyright (c) 2014 Moonlight Stream. All rights reserved.
//
//  Created by True砖家 on 2025.7.20.
//  Copyright © 2025 True砖家 @ Bilibili. All rights reserved.
//


#import "UIAppView.h"
#import "AppAssetManager.h"
#import "VoidLink-Swift.h"

static const float REFRESH_CYCLE = 1.0f;

@implementation UIAppView {
    TemporaryApp* _app;
    UILabel* _appLabel;
    UIImageView* _appOverlay;
    UIImageView* _appImage;
    NSCache* _artCache;
    id<AppCallback> _callback;
}

static UIImage* noImage;

- (id) initWithApp:(TemporaryApp*)app cache:(NSCache*)cache andCallback:(id<AppCallback>)callback {
    self = [super init];
    _app = app;
    _callback = callback;
    _artCache = cache;
    
    self.layer.cornerRadius = 16;
    self.clipsToBounds = YES;
    
    // Cache the NoAppImage ourselves to avoid
    // having to load it each time
    if (noImage == nil) {
        noImage = [UIImage imageNamed:@"NoAppImage"];
    }
        
#if TARGET_OS_TV
    self.frame = CGRectMake(0, 0, 200, 265);
#else
    self.frame = CGRectMake(0, 0, 150, 200);
#endif
    
    [self setAlpha:app.hidden ? 0.4 : 1.0];
    
    _appImage = [[UIImageView alloc] initWithFrame:self.frame];
    [_appImage setImage:noImage];
    
    [self addSubview:_appImage];
    
    // Use UIContextMenuInteraction on iOS 13.0+ and a standard UILongPressGestureRecognizer
    // for tvOS devices and iOS prior to 13.0.
#if !TARGET_OS_TV
    if (@available(iOS 13.0, *)) {
        UIContextMenuInteraction* rightClickInteraction = [[UIContextMenuInteraction alloc] initWithDelegate:self];
        [self addInteraction:rightClickInteraction];
    }
    else
#endif
    {
        UILongPressGestureRecognizer* longPressRecognizer = [[UILongPressGestureRecognizer alloc] initWithTarget:self action:@selector(appLongClicked:)];
        [self addGestureRecognizer:longPressRecognizer];
    }
    
    [self addTarget:self action:@selector(appClicked:) forControlEvents:UIControlEventPrimaryActionTriggered];
    
    [self addTarget:self action:@selector(buttonSelected:) forControlEvents:UIControlEventTouchDown];
    [self addTarget:self action:@selector(buttonDeselected:) forControlEvents:UIControlEventTouchUpInside | UIControlEventTouchCancel | UIControlEventTouchDragExit];
    
#if TARGET_OS_TV
    _appImage.adjustsImageWhenAncestorFocused = YES;
#else
    // Rasterizing the cell layer increases rendering performance by quite a bit
    // but we want it unrasterized for tvOS where it must be scaled.
    self.layer.shouldRasterize = NO;
    // self.layer.rasterizationScale = [UIScreen mainScreen].scale;
    
    if (@available(iOS 13.4.1, *)) {
        // Allow the button style to change when moused over
        self.pointerInteractionEnabled = YES;
    }
#endif
    
    [self updateAppImage];
    
    return self;
}

- (void)didMoveToSuperview {
    // Start our update loop when we are added to our cell
    if (self.superview != nil) {
        [self updateLoop];
    }
}

- (void) appClicked:(UIView *)view {
    [_callback appClicked:_app view:view];
}

- (void) appLongClicked:(UILongPressGestureRecognizer*)gesture {
    if (gesture.state == UIGestureRecognizerStateBegan) {
        [_callback appLongClicked:_app view:self];
    }
}

#if !TARGET_OS_TV
- (UIContextMenuConfiguration *)contextMenuInteraction:(UIContextMenuInteraction *)interaction
                        configurationForMenuAtLocation:(CGPoint)location {
    // We don't want to trigger the primary action at this point, so cancel
    // tracking touch on this view now. This will also have the (intended)
    // effect of removing the touch highlight on this view.
    [self cancelTrackingWithEvent:nil];
    
    [_callback appLongClicked:_app view:self];
    return nil;
}
#endif

- (bool)isCurrentApp{
    return [_app.id isEqualToString:_app.host.currentGame];
}

- (void) updateAppImage {

    if (_appOverlay != nil) {
        [_appOverlay removeFromSuperview];
        _appOverlay = nil;
    }
    if (_appLabel != nil) {
        [_appLabel removeFromSuperview];
        _appLabel = nil;
    }
    
    BOOL noAppImage = false;
    
    // First check the memory cache
    //UIImage* appImage = [_artCache objectForKey:_app];
    UIImage* appImage = nil;
    // NSLog(@"appImage instance: %lu", (uintptr_t)appImage);
    if (appImage == nil) {
        // Next try to load from the on disk cache
        appImage = [UIImage imageWithContentsOfFile:[AppAssetManager boxArtPathForApp:_app]];
        if (appImage != nil) {
            [_artCache setObject:appImage forKey:_app];
        }
    }
    
    
    // [_artCache setObject:appImage forKey:_app];

    if (appImage != nil) {
        // This size of image might be blank image received from GameStream.
        
        // TODO: Improve no-app image detection
        if (!(appImage.size.width == 130.f && appImage.size.height == 180.f) && // GFE 2.0
            !(appImage.size.width == 628.f && appImage.size.height == 888.f)){
        
        // if(true){ // GFE 3.0
            [_appImage setImage:appImage];
        } else {
            noAppImage = true;
        }
    } else {
        noAppImage = true;
    }
    
    
    if([self isCurrentApp]) {
        // Only create the app overlay if needed
        
        if (@available(iOS 13.0, *)) {
            UIImageSymbolConfiguration *config = [UIImageSymbolConfiguration configurationWithPointSize:23];
            UIImageView* playIcon = [[UIImageView alloc] initWithImage:[[UIImage systemImageNamed:@"play.circle.fill" withConfiguration:config] imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate]];
            //playIcon.tintColor = [ThemeManager.widgetBackgroundColor colorWithAlphaComponent:0.85];
            playIcon.tintColor = [[UIColor blackColor] colorWithAlphaComponent:0.55];

            _appOverlay = playIcon;
        } else {
            UIImageView* playIcon = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"play.circle.fill"]];
            playIcon.tintColor = [[UIColor blackColor] colorWithAlphaComponent:0.55];
            _appOverlay = playIcon;

        }

        
        //_appOverlay.layer.shadowColor = [UIColor blackColor].CGColor;
        //_appOverlay.layer.shadowOffset = CGSizeMake(1, 1);
        //_appOverlay.layer.shadowOpacity = 1;
        //_appOverlay.layer.shadowRadius = 1.3;
        //_appOverlay.contentMode = UIViewContentModeScaleAspectFit;
    }
    else if(noAppImage){
        if (@available(iOS 13.0, *)) {
            UIImageSymbolConfiguration *config = [UIImageSymbolConfiguration configurationWithPointSize:70];
            UIImage* appIconImage = [[UIImage imageNamed:@"icon-pc-app"] imageWithConfiguration:config];
            UIImageView* appIcon = [[UIImageView alloc] initWithImage:[appIconImage imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate]];
            
            //layIcon.tintColor = [ThemeManager.widgetBackgroundColor colorWithAlphaComponent:0.85];
            appIcon.tintColor = [[UIColor blackColor] colorWithAlphaComponent:0.5];

            _appOverlay = appIcon;
        }
    }
    
    if(true) {
        _appLabel = [[UILabel alloc] init];
        _appLabel.backgroundColor = [[UIColor blackColor] colorWithAlphaComponent:0.55];
        [_appLabel setTextColor:[[UIColor whiteColor] colorWithAlphaComponent:1]];
        //_appLabel.shadowColor = [UIColor blackColor];
        [_appLabel setText:[_app.name isEqualToString:@"Steam Big Picture"] ? @"Steam" : _app.name];
        [_appLabel setFont:[UIFont systemFontOfSize:15]];
        [_appLabel setBaselineAdjustment:UIBaselineAdjustmentAlignCenters];
        [_appLabel setTextAlignment:NSTextAlignmentCenter];
        [_appLabel setLineBreakMode:NSLineBreakByWordWrapping];
        [_appLabel setNumberOfLines:2];
        _appLabel.lineBreakMode = NSLineBreakByTruncatingTail;
    }
    
    [self positionSubviews];
    
#if TARGET_OS_TV
    [_appImage.overlayContentView addSubview:_appLabel];
    [_appImage.overlayContentView addSubview:_appOverlay];
#else
    [self addSubview:_appOverlay];
    [self addSubview:_appLabel];
#endif
}

- (void) buttonSelected:(id)sender {
    _appImage.layer.opacity = 0.5f;
}
- (void) buttonDeselected:(id)sender {
    _appImage.layer.opacity = [self isCurrentApp] ? 0.75 : 1.0f;
}

- (void) positionSubviews {
   // CGFloat padding = 28;
    CGFloat verticalPadding = 10;
    CGSize frameSize = _appImage.frame.size;
   //  CGPoint center = _appImage.center;
    
    [_appLabel setFrame:CGRectMake(0, frameSize.height-43, frameSize.width, 43)];
    //[_appLabel setFrame:CGRectMake(0, 0, 30, 12)];

    
        if (_appOverlay != nil) {
            _appOverlay.frame = [self isCurrentApp] ? CGRectMake(0, 0, frameSize.width / 2.39, frameSize.width / 2.39) : CGRectMake(0, 0, frameSize.width / 2, frameSize.width / 2);
            _appOverlay.center = CGPointMake(frameSize.width/2,  frameSize.height/2 - 2 * verticalPadding);
            if([self isCurrentApp]) _appImage.layer.opacity = 0.75f;

            //[_appLabel setFrame:CGRectMake(padding, _appOverlay.frame.size.height + padding, frameSize.width - 2 * padding, frameSize.height - _appOverlay.frame.size.height - 2 * padding)];
        }
}

- (void) updateLoop {
    // Stop immediately if the view has been detached
    if (self.superview == nil || ![self.updateLoopDelegate isInAppView]) {
        return;
    }
    
    NSLog(@"appview update loop %f", CACurrentMediaTime());
    
    // Update the app image if neccessary
    if ((_appOverlay != nil && ![_app.id isEqualToString:_app.host.currentGame]) ||
        (_appOverlay == nil && [_app.id isEqualToString:_app.host.currentGame])) {
        [self updateAppImage];
    }
    
    // Show no shadow for hidden apps. Because we adjust the opacity of the
    // cells for hidden apps, it makes them look bad when the shadow draws
    // through the app tile.
    // self.superview.layer.shadowOpacity = _app.hidden ? 0.0f : 0.5f;
    self.superview.layer.shadowOpacity = 0;

    // Update opacity if neccessary
    [self setAlpha:_app.hidden ? 0.4 : 1.0];
    
    // Queue the next refresh cycle
    [self performSelector:@selector(updateLoop) withObject:self afterDelay:REFRESH_CYCLE];
}

@end

File: VoidLink/UIAppView.swift
//
//  UIAppView.swift
//  Moonlight
//
//  Created by Diego Waxemberg on 10/22/14.
//  Copyright (c) 2014 Moonlight Stream. All rights reserved.
//
//  Created by True Zhuanjia on 2025.7.20.
//  Copyright 2025 True Zhuanjia @ Bilibili. All rights reserved.
//

import UIKit

private let refreshCycle: TimeInterval = 1.0

@objc protocol AppViewUpdateLoopDelegate: AnyObject {
    func isInAppView() -> Bool
}

@objc protocol AppCallback: AnyObject {
    @objc(appClicked:view:)
    func appClicked(_ app: TemporaryApp, view: UIView)

    @objc(appLongClicked:view:)
    func appLongClicked(_ app: TemporaryApp, view: UIView)
}

@objcMembers
final class UIAppView: UIButton {
    weak var updateLoopDelegate: AppViewUpdateLoopDelegate?

    private static var noImage: UIImage?

    let app: TemporaryApp
    fileprivate weak var callback: AppCallback?
    private let artCache: NSCache<AnyObject, AnyObject>
    private var appLabel: UILabel?
    private var appOverlay: UIImageView?
    private var appImage: UIImageView
    private var contextMenuDelegate: AnyObject?

    @objc(initWithApp:cache:andCallback:)
    init(app: TemporaryApp, cache: NSCache<AnyObject, AnyObject>, andCallback callback: AppCallback) {
        self.app = app
        self.callback = callback
        self.artCache = cache

        if UIAppView.noImage == nil {
            UIAppView.noImage = UIImage(named: "NoAppImage")
        }

        #if os(tvOS)
        let initialFrame = CGRect(x: 0, y: 0, width: 200, height: 265)
        #else
        let initialFrame = CGRect(x: 0, y: 0, width: 150, height: 200)
        #endif

        self.appImage = UIImageView(frame: initialFrame)

        super.init(frame: .zero)

        layer.cornerRadius = 16
        if #available(iOS 13.0, tvOS 13.0, *) {
            layer.cornerCurve = .continuous
        }
        clipsToBounds = true
        frame = initialFrame
        alpha = app.hidden ? 0.4 : 1.0

        appImage.image = UIAppView.noImage
        addSubview(appImage)

        #if !os(tvOS)
        if #available(iOS 13.0, *) {
            let delegate = UIAppViewContextMenuDelegate(appView: self)
            contextMenuDelegate = delegate
            let rightClickInteraction = UIContextMenuInteraction(delegate: delegate)
            addInteraction(rightClickInteraction)
        } else {
            let longPressRecognizer = UILongPressGestureRecognizer(target: self, action: #selector(appLongClicked(_:)))
            addGestureRecognizer(longPressRecognizer)
        }
        #else
        let longPressRecognizer = UILongPressGestureRecognizer(target: self, action: #selector(appLongClicked(_:)))
        addGestureRecognizer(longPressRecognizer)
        #endif

        addTarget(self, action: #selector(appClicked(_:)), for: .primaryActionTriggered)
        addTarget(self, action: #selector(buttonSelected(_:)), for: .touchDown)
        addTarget(self, action: #selector(buttonDeselected(_:)), for: [.touchUpInside, .touchCancel, .touchDragExit])

        #if os(tvOS)
        appImage.adjustsImageWhenAncestorFocused = true
        #else
        layer.shouldRasterize = false
        if #available(iOS 13.4, *) {
            isPointerInteractionEnabled = true
        }
        #endif

        updateAppImage()
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func didMoveToSuperview() {
        super.didMoveToSuperview()
        if superview != nil {
            updateLoop()
        }
    }

    @objc private func appClicked(_ view: UIView) {
        callback?.appClicked(app, view: view)
    }

    @objc private func appLongClicked(_ gesture: UILongPressGestureRecognizer) {
        if gesture.state == .began {
            callback?.appLongClicked(app, view: self)
        }
    }

    private var isCurrentApp: Bool {
        app.id == app.host?.currentGame
    }

    func updateAppImage() {
        appOverlay?.removeFromSuperview()
        appOverlay = nil
        appLabel?.removeFromSuperview()
        appLabel = nil

        var noAppImage = false
        var loadedAppImage: UIImage?

        if let boxArtPath = AppAssetManager.boxArtPath(for: app) {
            loadedAppImage = UIImage(contentsOfFile: boxArtPath)
            if let loadedAppImage {
                artCache.setObject(loadedAppImage, forKey: app)
            }
        }

        if let loadedAppImage {
            let isGFE2BlankImage = loadedAppImage.size.width == 130.0 && loadedAppImage.size.height == 180.0
            let isGFE3BlankImage = loadedAppImage.size.width == 628.0 && loadedAppImage.size.height == 888.0
            if !isGFE2BlankImage && !isGFE3BlankImage {
                appImage.image = loadedAppImage
            } else {
                noAppImage = true
            }
        } else {
            noAppImage = true
        }

        if isCurrentApp {
            if #available(iOS 13.0, tvOS 13.0, *) {
                let config = UIImage.SymbolConfiguration(pointSize: 23)
                let image = UIImage(systemName: "play.circle.fill", withConfiguration: config)?.withRenderingMode(.alwaysTemplate)
                let playIcon = UIImageView(image: image)
                playIcon.tintColor = UIColor.black.withAlphaComponent(0.55)
                appOverlay = playIcon
            } else {
                let playIcon = UIImageView(image: UIImage(named: "play.circle.fill"))
                playIcon.tintColor = UIColor.black.withAlphaComponent(0.55)
                appOverlay = playIcon
            }
        } else if noAppImage {
            if #available(iOS 13.0, tvOS 13.0, *) {
                let config = UIImage.SymbolConfiguration(pointSize: 70)
                let image = UIImage(named: "icon-pc-app")?.withConfiguration(config).withRenderingMode(.alwaysTemplate)
                let appIcon = UIImageView(image: image)
                appIcon.tintColor = UIColor.black.withAlphaComponent(0.5)
                appOverlay = appIcon
            }
        }

        let label = UILabel()
        label.backgroundColor = UIColor.black.withAlphaComponent(0.55)
        label.textColor = UIColor.white
        label.text = app.name == "Steam Big Picture" ? "Steam" : app.name
        label.font = UIFont.systemFont(ofSize: 15)
        label.baselineAdjustment = .alignCenters
        label.textAlignment = .center
        label.lineBreakMode = .byTruncatingTail
        label.numberOfLines = 2
        appLabel = label

        positionSubviews()

        #if os(tvOS)
        appImage.overlayContentView.addSubview(label)
        if let appOverlay {
            appImage.overlayContentView.addSubview(appOverlay)
        }
        #else
        if let appOverlay {
            addSubview(appOverlay)
        }
        addSubview(label)
        #endif
    }

    @objc private func buttonSelected(_ sender: Any) {
        appImage.layer.opacity = 0.5
    }

    @objc private func buttonDeselected(_ sender: Any) {
        appImage.layer.opacity = isCurrentApp ? 0.75 : 1.0
    }

    private func positionSubviews() {
        let verticalPadding: CGFloat = 10
        let frameSize = appImage.frame.size

        appLabel?.frame = CGRect(x: 0, y: frameSize.height - 43, width: frameSize.width, height: 43)

        guard let appOverlay else { return }
        let overlaySize = isCurrentApp ? frameSize.width / 2.39 : frameSize.width / 2.0
        appOverlay.frame = CGRect(x: 0, y: 0, width: overlaySize, height: overlaySize)
        appOverlay.center = CGPoint(x: frameSize.width / 2.0, y: frameSize.height / 2.0 - 2.0 * verticalPadding)

        if isCurrentApp {
            appImage.layer.opacity = 0.75
        }
    }

    @objc private func updateLoop() {
        guard superview != nil, updateLoopDelegate?.isInAppView() == true else {
            return
        }

        if (appOverlay != nil && !isCurrentApp) || (appOverlay == nil && isCurrentApp) {
            updateAppImage()
        }

        superview?.layer.shadowOpacity = 0
        alpha = app.hidden ? 0.4 : 1.0

        perform(#selector(updateLoop), with: self, afterDelay: refreshCycle)
    }
}

@available(iOS 13.0, *)
private final class UIAppViewContextMenuDelegate: NSObject, UIContextMenuInteractionDelegate {
    private weak var appView: UIAppView?

    init(appView: UIAppView) {
        self.appView = appView
    }

    func contextMenuInteraction(_ interaction: UIContextMenuInteraction,
                                configurationForMenuAtLocation location: CGPoint) -> UIContextMenuConfiguration? {
        guard let appView else { return nil }
        appView.cancelTracking(with: nil)
        appView.callback?.appLongClicked(appView.app, view: appView)
        return nil
    }
}
